### ISSUE TITLE
Fine-tuning multiple hard negatives in a single training instance

### ISSUE THREAD

**Issue by thakur-nandan** [lightonai/pylate#117]
Hi Pylate authors, not an issue, but more of a question. Thank you for this package!

Is it possible to fine-tune ColBERT style models in either contrastive training or knowledge distillation using a single training instance with multiple hard negatives or HNs (e.g., one positive and 7 negatives), which is provided in a Tevatron format -- <query, positive, HN1, HN2,..., HN7>? I only saw examples with triplet-style pairs containing <query, positive, HN1>.

Existing training datasets such as MSMARCO contain multiple HN during training. A naive way is to expand each hard negative with a positive pair as a separate triplet, however, this increases the training dataset size by K times (where K is the number of HNs). 

I was wondering if there was an easier way to do this in PyLate, if possible? 

Thanks,
Nandan

**Comment by NohTow** [comment]
Hey,
Actually, you do not have to create multiple instances: PyLate already handle multiple negatives per sample!
It should be better documented, but essentially you can see [there](https://github.com/lightonai/pylate/blob/86b969185aec140d34819335be5bfd41983a256d/pylate/losses/contrastive.py#L150) that we are computing the embeddings for every column.
Then, the labels are just [torch.arange(batch_size)](https://github.com/lightonai/pylate/blob/86b969185aec140d34819335be5bfd41983a256d/pylate/losses/contrastive.py#L163) and given how we [concatenate the scores](https://github.com/lightonai/pylate/blob/86b969185aec140d34819335be5bfd41983a256d/pylate/losses/contrastive.py#L185), the positives matches are actually just the first and second set of elements (queries and positives), everything else are negatives.
This means that you can put as many negatives columns as you want, you just have to put in the order (query, pos, neg_1, neg_2, ..., neg_n).
Hope it helps!

**Comment by NohTow** [comment]
You're welcome!
Cc @raphaelsty would be cool to add this to the documentation, when we add the gathering and gradcache documentation
